sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala

```diff
@@ -0,0 +1,194 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.pipelines
+
+import java.util.concurrent.{BlockingQueue, Executors, ExecutorService, LinkedBlockingQueue}
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.util.control.NonFatal
+
+import com.google.protobuf.{Timestamp => ProtoTimestamp}
+import io.grpc.stub.StreamObserver
+
+import org.apache.spark.connect.proto
+import org.apache.spark.connect.proto.ExecutePlanResponse
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connect.service.SessionHolder
+import org.apache.spark.sql.pipelines.logging.PipelineEvent
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * Handles sending pipeline events to the client in a background thread using ExecutorService.
+ * This prevents pipeline execution from blocking on response observer operations and provides
+ * better error isolation between pipeline execution and client communication.
+ *
+ * Based on the executor pattern used in DataPlaneEventLog for better thread management.
+ */
+class PipelineEventSender(
+    responseObserver: StreamObserver[ExecutePlanResponse],
+    sessionHolder: SessionHolder)
+    extends Logging {
+
+  // Thread-safe queue for buffering events
+  private val eventQueue: BlockingQueue[PipelineEventMessage] = new LinkedBlockingQueue()
+
+  // ExecutorService for background event processing
+  private val executor: ExecutorService = {
+    val threadFactory = ThreadUtils
+      .namedThreadFactory(s"pipeline-event-sender-${sessionHolder.sessionId}")
+    Executors.newSingleThreadExecutor(threadFactory)
+  }
+
+  // state tracking
+  private val isShutdown = new AtomicBoolean(false)
+  private val isStarted = new AtomicBoolean(false)
```

Comment:
Let's document how/when these state transitions happen and what these flags are used for